### PR TITLE
Arcadia

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Non-significant (changelog entry is not required)

### CODE REVIEW

File: base/common/StringRef.cpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#include "StringRef.h"
+
+#include <city.h>
+
+#if !defined(ARCADIA_BUILD)
+using namespace CityHash_v1_0_2;
+#endif
+
+size_t StringRefHash64::operator()(StringRef x) const
```
Comment (alexey-milovidov) [inline base/common/StringRef.cpp]: I'm afraid that it cannot inline.

File: base/common/StringRef.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+{
+    return CityHash64(x.data, x.size);
+}
+
+#if defined(__SSE4_2__)
+
+UInt64 hashLen16(UInt64 u, UInt64 v)
+{
+    return Hash128to64(uint128(u, v));
```
Comment (alexey-milovidov) [inline base/common/StringRef.cpp]: And here I'm afraid even more.

File: base/common/ya.make [hunk 3]
```diff
@@ -1,12 +1,53 @@
 LIBRARY()
 
+ADDINCL(
+    # public
+    GLOBAL clickhouse/base
+    GLOBAL clickhouse/dbms/src
+    GLOBAL util/digest # for <city.h>
```
Comment (alexey-milovidov) [inline base/common/ya.make]: Need to double-check that we are using fixed version of CityHash64 even in Arcadia.

File: base/daemon/BaseDaemon.cpp [hunk 4]
```diff
@@ -222,7 +225,11 @@ class SignalListener : public Poco::Runnable
 
                 /// This allows to receive more signals if failure happens inside onFault function.
                 /// Example: segfault while symbolizing stack trace.
+#if __cplusplus <= 201703L
```
Comment (alexey-milovidov) [inline base/daemon/BaseDaemon.cpp]: Need to enable C++20 in Arcadia.
The C++ committee does not have anything against C++20 (they didn't answer our email).

File: dbms/programs/server/HTTPHandler.cpp [hunk 5]
```diff
@@ -15,7 +15,11 @@
 #include <common/getFQDNOrHostName.h>
 #include <Common/CurrentThread.h>
 #include <Common/setThreadName.h>
-#include <Common/config.h>
+
+#if !defined(ARCADIA_BUILD)
+#    include <Common/config.h>
```
Comment (alexey-milovidov) [inline dbms/programs/server/HTTPHandler.cpp]: Looks like it is unused here.

File: dbms/src/Access/AccessRightsContext.cpp [hunk 6]
```diff
@@ -157,9 +157,9 @@ void AccessRightsContext::setUser(const UserPtr & user_) const
         current_roles.reserve(params.current_roles.size());
         for (const auto & id : params.current_roles)
         {
-            if (user->granted_roles.contains(id))
+            if (user->granted_roles.find(id) != user->granted_roles.end())
```
Comment (alexey-milovidov) [inline dbms/src/Access/AccessRightsContext.cpp]: I don't appreciate this.

File: dbms/src/AggregateFunctions/AggregateFunctionUniq.h [hunk 7]
```diff
@@ -177,7 +177,10 @@ struct OneAdder
             else
             {
                 StringRef value = column.getDataAt(row_num);
-                data.set.insert(CityHash_v1_0_2::CityHash64(value.data, value.size));
+#if !defined(ARCADIA_BUILD)
+                using namespace CityHash_v1_0_2;
+#endif
+                data.set.insert(CityHash64(value.data, value.size));
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionUniq.h]: Previous code was more clear.
Why do we need that?

File: dbms/src/Columns/ColumnVector.cpp [hunk 8]
```diff
@@ -176,10 +179,14 @@ void ColumnVector<T>::getPermutation(bool reverse, size_t limit, int nan_directi
         for (size_t i = 0; i < s; ++i)
             res[i] = i;
 
+#if !defined(ARCADIA_BUILD)
         if (reverse)
             pdqsort(res.begin(), res.end(), greater(*this, nan_direction_hint));
         else
             pdqsort(res.begin(), res.end(), less(*this, nan_direction_hint));
+#else
+        /// TODO: implement this
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnVector.cpp]: Tests should fail.

File: dbms/src/Access/AccessRightsContext.cpp (hunk 6 above)
Comment (abyss7) [reply dbms/src/Access/AccessRightsContext.cpp]: Replaced with a `count()` for now. Don't think, that `#ifdef`s will be better.

**Comment by abyss7** [comment]
Closed in favour of #10307

**Pull request closed without merging**